tensorzero-internal/src/inference/providers/helpers.rs

```diff
@@ -12,20 +13,100 @@ pub fn inject_extra_body(
     config: Option<&ExtraBodyConfig>,
     body: &mut serde_json::Value,
 ) -> Result<(), Error> {
-    let Some(body_map) = body.as_object_mut() else {
+    if !body.is_object() {
         return Err(Error::new(ErrorDetails::Serialization {
             message: "Body is not a map".to_string(),
         }));
-    };
+    }
     if let Some(extra_body) = config.as_ref() {
-        for (key, value) in extra_body.data.iter() {
-            // We intentionally overwrite any existing keys in the body.
-            body_map.insert(key.to_string(), value.clone());
+        for replacement in &extra_body.data {
+            write_json_pointer_with_parent_creation(
+                body,
+                &replacement.pointer,
+                replacement.value.clone(),
+            )?;
         }
     }
     Ok(())
 }
 
+// Copied from serde_json (MIT-licensed): https://github.com/serde-rs/json/blob/400eaa977f1f0a1c9ad5e35d634ed2226bf1218c/src/value/mod.rs#L259
+fn parse_index(s: &str) -> Option<usize> {
+    if s.starts_with('+') || (s.starts_with('0') && s.len() != 1) {
+        return None;
+    }
+    s.parse().ok()
+}
```

Comment:
can you add a comment explaining the semantics here and how it's used in the code below?